test/integration/volume/attach_detach_test.go

```diff
@@ -361,6 +549,69 @@ func TestPodUpdateWithKeepTerminatedPodVolumes(t *testing.T) {
 	waitForPodFuncInDSWP(t, ctrl.GetDesiredStateOfWorld(), 20*time.Second, "expected non-zero pods in dsw if KeepTerminatedPodVolumesAnnotation is set", 1)
 }
 
+func waitForMetric(t *testing.T, m basemetric.CounterMetric, expectedCount float64, identifier string) {
+	if err := wait.Poll(100*time.Millisecond, 60*time.Second, func() (bool, error) {
+		gotCount, err := metricstestutil.GetCounterMetricValue(m)
+		fmt.Println(gotCount)
+		if err != nil {
+			t.Fatal(err, identifier)
+		}
+		if gotCount >= expectedCount {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err, identifier)
+	}
+}
+
+func waitForNodeToBeTainted(t *testing.T, testingClient *clientset.Clientset, nodeName, taintKey string) {
+	if err := wait.Poll(100*time.Millisecond, 60*time.Second, func() (bool, error) {
+		node, err := testingClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == taintKey {
+				return true, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForNodeToBeNotReady(t *testing.T, testingClient *clientset.Clientset, nodeName string) {
+	if err := wait.Poll(100*time.Millisecond, 60*time.Second, func() (bool, error) {
+		node, err := testingClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !nodeutil.IsNodeReady(node) {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForPodDeletionTimeStampToSet(t *testing.T, testingClient *clientset.Clientset, podName, podNamespace string) {
+	if err := wait.Poll(100*time.Millisecond, 60*time.Second, func() (bool, error) {
+		pod, err := testingClient.CoreV1().Pods(podNamespace).Get(context.TODO(), podName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		if pod.DeletionTimestamp != nil {
+			return true, nil
+		}
```

Comment:
this will be racy if the pod does not have a finalizer because can be deleted and return a 404 Not found error, no?

Reply:
The pod in question has been set a grace time period of 300 seconds in this test to allow enough time. So this should be fine. Let me know if you still have a different opinion.

Reply:
sounds good, thanks for clarifying